File: 3605.minimum-stability-factor-of-array.java
#
# @lc app=leetcode id=3605 lang=java
#
# [3605] Minimum Stability Factor of Array
#

# @lc code=start
import java.util.*;
class Solution {
    public int minStable(int[] nums, int maxC) {
        int n = nums.length;
        int left = 0, right = n;
        while (left < right) {
            int mid = (left + right) / 2;
            if (canReduce(nums, maxC, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
    // Check if it's possible to make all stable subarrays have length <= k after maxC changes
    private boolean canReduce(int[] nums, int maxC, int k) {
        int n = nums.length;
        int changes = 0;
        // For every possible subarray of length k+1, check if it's stable
        // We can optimize by only considering windows with GCD >= 2
        for (int i = 0; i + k < n; ) {
            int g = nums[i];
            int j = i + 1;
            while (j <= i + k && j < n) {
                g = gcd(g, nums[j]);
                j++;
            }
            if (g >= 2) {
                changes++;
                // To break this window, we can skip to the end of the window
                i = j;
            } else {
                i++;
            }
            if (changes > maxC) return false;
        }
        return true;
    }
    private int gcd(int a, int b) {
        while (b != 0) {
            int t = b;
            b = a % b;
            a = t;
        }
        return a;
    }
}
# @lc code=end

File: 3605.minimum-stability-factor-of-array.go
#
# @lc app=leetcode id=3605 lang=golang
#
# [3605] Minimum Stability Factor of Array
#

# @lc code=start
import (
    "math"
)

func gcd(a, b int) int {
    for b != 0 {
        a, b = b, a % b
    }
    return a
}

// (Implementation outline, not full code)
func minStable(nums []int, maxC int) int {
    n := len(nums)
    left, right := 1, n
    answer := n
    
    // Precompute sparse table or segment tree for GCD queries
    // ...

    // Helper: Checks if we can eliminate all stable subarrays of length >= k with <= maxC modifications
    isPossible := func(k int) bool {
        // 1. For all windows of length k, get GCD efficiently
        // 2. For those with GCD >= 2, treat as intervals [i, i+k-1]
        // 3. Use greedy interval covering: sort intervals by end, greedily select covering points
        // 4. Count minimal number of points required, compare to maxC
        // Return true if feasible, false otherwise
        return true // placeholder
    }
    
    for left <= right {
        mid := (left + right) / 2
        if isPossible(mid) {
            answer = mid
            right = mid - 1
        } else {
            left = mid + 1
        }
    }
    return answer
}
# @lc code=end